### PR TITLE
Fix #4315 - Wrap E+ FuelFactors object

### LINKED ISSUE #4315: Wrap E+ FuelFactors object
Enhancement Request
--------------
The [EnergyPlus FuelFactors object](https://bigladdersoftware.com/epx/docs/9-5/input-output-reference/input-for-output.html#fuelfactors) is not available in the OS SDK. We have a lot of interest in using this for calculating carbon metrics (similar to TDV) in the near-ish future and it would be nice to not have to resort to an EnergyPlus measure.

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

Fix #4315 - Wrap E+ FuelFactors object

This PR adds three new objects:
* OutputEnvironmentalImpactFactors
* EnvironmentalImpactFactors: this is a unique model object
* FuelFactors

OutputEnvironmentalImpactFactors is the one that triggers FT of the two others (in E+, it's the object that actually enables the pollution reporting module).
OutputEnvironmentalImpactFactors also checks that there is at least ONE FuelFactors, as you must provide one FuelFactors per each fuel you actually use in your model.

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [x] Model API Changes / Additions
 - [x] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [x] Model API methods are tested (in `src/model/test`)
 - [x] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [x] If a new object or method, added a test in NREL/OpenStudio-resources: https://github.com/NREL/OpenStudio-resources/pull/139
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [x] All new and existing tests passes

**Labels:**

 - [x] If change to an IDD file, add the label `IDDChange`
 - [x] If breaking existing API, add the label `APIChange`
 - [x] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: resources/model/OpenStudio.idd [hunk 1]
```diff
@@ -33367,6 +33367,211 @@ OS:Meter:CustomDecrement,
...
+       \type choice
+       \key Electricity
+       \key NaturalGas
+       \key FuelOilNo1
+       \key FuelOilNo2
+       \key Coal
+       \key Gasoline
+       \key Propane
+       \key Diesel
+       \key OtherFuel1
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:33437]: I **removed**   A2,   \field Units of Measure and   N1,   \field Energy per Unit Factor

They are NOT used in the E+ source code.

File: src/energyplus/ForwardTranslator.cpp [hunk 2]
```diff
@@ -3061,6 +3076,12 @@ namespace energyplus {
     result.push_back(IddObjectType::OS_Output_DebuggingData);
     result.push_back(IddObjectType::OS_Output_Diagnostics);
     result.push_back(IddObjectType::OS_Output_JSON);
+
+    // Note: we just always translate Output:EnvironmentalImpactFactors, and in there (it exists), then trigger translatation of the two others
+    result.push_back(IddObjectType::OS_Output_EnvironmentalImpactFactors);
+    // result.push_back(IddObjectType::OS_EnvironmentalImpactFactors);
+    // result.push_back(IddObjectType::OS_FuelFactors);
+
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator.cpp:3084]: FYI

File: src/energyplus/ForwardTranslator/ForwardTranslateOutputEnvironmentalImpactFactors.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    idfObject.setString(Output_EnvironmentalImpactFactorsFields::ReportingFrequency, modelObject.reportingFrequency());
+
+    for (auto& fuelFactor : fuelFactors) {
+      translateAndMapModelObject(fuelFactor);
+    }
+
+    for (auto& environmentalImpactFactor : modelObject.model().getConcreteModelObjects<EnvironmentalImpactFactors>()) {
+      translateAndMapModelObject(environmentalImpactFactor);
+    }
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateOutputEnvironmentalImpactFactors.cpp:73]: FYI, FT logic

File: src/model/test/FuelFactors_GTest.cpp [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+  // Bad Value
+  EXPECT_FALSE(fuelFactors.setNuclearLowLevelEmissionFactor(-10.0));
+  EXPECT_EQ(0.1, fuelFactors.nuclearLowLevelEmissionFactor());
+
+  // Nuclear Low Level Emission Factor Schedule Name: Optional Object
+  ScheduleConstant sch17(m);
+  EXPECT_TRUE(fuelFactors.setNuclearLowLevelEmissionFactorSchedule(sch17));
+  ASSERT_TRUE(fuelFactors.nuclearLowLevelEmissionFactorSchedule());
+  EXPECT_EQ(sch17, fuelFactors.nuclearLowLevelEmissionFactorSchedule().get());
+}
```
Comment (tijcolem) [inline src/model/test/FuelFactors_GTest.cpp:272]: Wow, seems like a nice comprehensive set of test cases!

**Comment by tijcolem** [comment]
@jmarrec Great work!  Seems like added test cases have great coverage. 

@shorowit Would you be able to test out on the installers posted in the comments?

**Review by tijcolem — APPROVED** [review]

**Comment by shorowit** [comment]
I didn't do comprehensive testing, but the few simple-ish tests I did look good.

One thing I noticed is that you can create multiple FuelFactors objects with the same fuel type (e.g., "electricity"). The OSM ends up with two objects and the EnergyPlus simulation runs but gives the warning `** Warning ** FuelFactors: ELECTRICITY already entered. Previous entry will be used.`. Not sure if OpenStudio needs to be more helpful here, just mentioning the current behavior.

**Comment by jmarrec** [comment]
> I didn't do comprehensive testing, but the few simple-ish tests I did look good.
> 
> One thing I noticed is that you can create multiple FuelFactors objects with the same fuel type (e.g., "electricity"). The OSM ends up with two objects and the EnergyPlus simulation runs but gives the warning `** Warning ** FuelFactors: ELECTRICITY already entered. Previous entry will be used.`. Not sure if OpenStudio needs to be more helpful here, just mentioning the current behavior.

Yeah I'm aware. I chose not to do it, as I doubt throwing such a warning in the FT would have more impact than letting E+ do it. We can do it if you want though. Here: https://github.com/NREL/OpenStudio/blob/c5d74654c70cfaf0b1d716d47155f026f6349349/src/energyplus/ForwardTranslator/ForwardTranslateFuelFactors.cpp#L54-L55

```c++
    #include <algorithm>
    [...]
    // Existing Fuel Resource Name
    std::string fuelResourceName = modelObject.existingFuelResourceName();
    idfObject.setString(FuelFactorsFields::ExistingFuelResourceName, fuelResourceName);
    auto ffs = modelObject.model().getConcreteModelObjects<FuelFactors>();
    if (std::count_if(ffs.cbegin(), ffs.cend(), [&fuelResourceName](const auto& f) { return f.existingFuelResourceName() == fuelResourceName; }) > 1) {
      LOG(Warn, "FuelFactors: fuel '" << fuelResourceName << "' already entered. EnergyPlus will use the first one it encounters");
    }
```

Moving this check at model time will be annoying: you have to modify the Constructor to take the `Fuel Resource Name` as an argument and potentially throw in the Ctor, and do a check in `setExistingFuelResource` as well. I'd consider it if it was an object that I thought would be commonly used, which isn't the case here.

Let me know if you think a check in the FT is warranted @shorowit

**Comment by shorowit** [comment]
I don't think it's needed, no. If EnergyPlus threw an error (which it really should be doing, IMO) instead of a warning, then I would have a different opinion here, but the current behavior seems fine.

**Comment by tijcolem** [comment]
@shorowit Thanks for testing. Sounds like this is good to go and ready to drop in.

**Review by shorowit — APPROVED** [review]

**Pull request merged**